contents/tutorials/monitor-aws-bedrock-calls.md

```diff
@@ -0,0 +1,354 @@
+    const output_cost_in_dollars = generation_token_count * token_output_cost;
+    const total_cost_in_dollars = input_cost_in_dollars + output_cost_in_dollars;
+
+    posthog.capture({
+      distinctId: email,
+      event: 'bedrock_completion',
+      properties: {
+        // ...existing properties
+        input_cost_in_dollars,
+        output_cost_in_dollars,
+        total_cost_in_dollars,
+      }
+    })
+  }
+```
+
+### c. API response time
+
+API responses can take long, especially for longer outputs, so it's useful to keep an eye on this. To do this, we track the request start and end times and calculate the total time. Then, we include the response time in the event properties:
+
+```js file=src/app/api/generate-recipe/route.js
+// your existing code
+
+  try { 
+    // ... existing code
+
+    const startTime = performance.now(); // add just before the API request
+
+    const response = await client.send(command);
+    
+    const endTime = performance.now(); // add just after the API request
+    const responseTime = endTime - startTime;
+
+    // ... existing code
+
+    posthog.capture({
+      distinctId: email,
+      event: 'bedrock_completion',
+      properties: {
+        response_time_in_ms: responseTime
+        // ...any other properties you're capturing
+      }
+    })
+  }
+```
+
+### d. Errors
+
+It's not uncommon for generative AI requests to fail and it's important to track these errors. To do this, we capture a `bedrock_error` event in the `catch` block of our code:
+
+```js file=src/app/api/generate-recipe/route.js
+// your existing code
+  } catch (error) {
+    console.error('Error:', error);
+
+    posthog.capture({
+      distinctId: email,
+      event: 'bedrock_error',
+      properties: {
+        error_message: error.message,
+        error_name: error.name,
+        error_code: error.$metadata?.httpStatusCode,
+        error_type: error.__type,
+        error_requestId: error.$metadata?.requestId,
+        error_stack: error.stack,
+        model_id: modelId,
+        prompt,
+      }
+    });
+
+    return new Response(JSON.stringify({ error: 'Failed to generate recipe' }), {
+      status: 500,
+      headers: { 'Content-Type': 'application/json' },
+    });
+  } finally {
+    // rest of your code
+  }
+```
+
+## 4. Create insights
+
+Now that we're capturing events, we can create [insights](/docs/product-analytics/insights) in PostHog to visualize our data. Below are five examples of useful metrics to track:
+
+### a. Generation count
+
+**What it is:** The total number of successful requests to your model.
+
+**Why it's useful:** Helps assess the workload and demand placed on your models, which directly impacts costs and performance.
+
+**How to set it up:** 
+
+1. Go the [Product analytics tab](https://us.posthog.com/insights) and click **+ New insight**
+2. Set the event to `bedrock_completion`
+3. Ensure the second dropdown shows **Total count**
+4. Press **Save**
```

Comment:
```suggestion
Now that we're capturing events, we can create [insights](/docs/product-analytics/insights) in PostHog to visualize our data. Below are five examples of useful metrics to track. Each of these starts by going to the [Product analytics tab](https://us.posthog.com/insights) and clicking **+ New insight**.

### a. Generation count

**What it is:** The total number of successful requests to your model.

**Why it's useful:** Helps assess the workload and demand placed on your models, which directly impacts costs and performance.

**How to set it up:** 

1. Set the event to `bedrock_completion`
2. Ensure the second dropdown shows **Total count**
3. Press **Save**
```

Reply:
Maybe add "Set the event to `bedrock_completion`" to the intro paragraph too because you repeat that requirement in each of the steps.